### PR TITLE
Added Data manipulator/Keys for Disabled Slots in an ArmorStand.

### LINKED ISSUE #1837: No Key for the armor stand property DisabledSlots
(Empty issue description)

---
**Edit:**  (ryantheleach)
https://minecraft.gamepedia.com/Armor_Stand#Entity_data|

https://bugs.mojang.com/browse/MC-77737

https://www.reddit.com/r/Minecraft/comments/2czu1p/armorstand_disabledslots_flag_system/

### PR DESCRIPTION
**Pull request by tcmal**
Heyo,

I'm not sure if I've done everything needed to register a new DataManipulator so I've done what seemed relevant.

DisabledSlotsData (DISABLED_SLOTS_DATA in CatalogEntityData) has methods:

  - `takingDisabled()` (`ARMOR_STAND_TAKING_DISABLED`)
  - `placingDisabled()` (`ARMOR_STAND_PLACING_DISABLED`)

which each return `Map<EquipmentType, Boolean>`. The boolean for each slot is if that action is disabled (ie false allows access).

Fixes #1837

### CODE REVIEW

File: src/main/java/org/spongepowered/api/data/key/Keys.java [hunk 1]
```diff
@@ -278,6 +280,18 @@
      */
     public static final Key<Value<Boolean>> ARMOR_STAND_MARKER = DummyObjectProvider.createExtendedFor(Key.class,"ARMOR_STAND_MARKER");
 
+	/**
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/data/key/Keys.java]: Replace `<TAB>` with 4 spaces please.

**Comment by Zidane** [comment]
This can actually target stable-7.

File: src/main/java/org/spongepowered/api/data/key/Keys.java [hunk 2]
```diff
@@ -278,6 +280,18 @@
...
+	 * Represents the {@link Key} for whether players are prevented from taking
+	 * items from an equipment slot on an {@link ArmorStand}
+	 */
+	public static final Key<Value<Map<EquipmentType, Boolean>>> ARMOR_STAND_TAKING_DISABLED = DummyObjectProvider.createExtendedFor(Key.class, "ARMOR_STAND_TAKING_DISABLED");
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/data/key/Keys.java]: We have a `Map` based `Value` subtype: [`MapValue`](https://github.com/SpongePowered/SpongeAPI/blob/1e966facc18ff0c5e9456b4ec0e9bf70f386633d/src/main/java/org/spongepowered/api/data/value/mutable/MapValue.java#L45), it's recommended to use that type in the declaration.
```suggestion
+    public static final Key<MapValue<EquipmentType, Boolean>> ARMOR_STAND_TAKING_DISABLED = DummyObjectProvider.createExtendedFor(Key.class, "ARMOR_STAND_TAKING_DISABLED");
```

File: src/main/java/org/spongepowered/api/data/manipulator/catalog/CatalogEntityData.java [hunk 3]
```diff
@@ -103,6 +103,13 @@
      * will calm down after a certain amount of time has passed.
      */
     public static final Class<AngerableData> ANGERABLE_DATA = AngerableData.class;
+    
+	/**
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/data/manipulator/catalog/CatalogEntityData.java]: Same here regarding tabs.

File: src/main/java/org/spongepowered/api/data/manipulator/immutable/entity/ImmutableDisabledSlotsData.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import org.spongepowered.api.item.inventory.equipment.EquipmentType;
+
+/**
+ * A {@link ImmutableDataManipulator} for the disabled slots in an
+ * {@link ArmorStand}.
+ * 
+ * A slot can have taking disabled, placing disabled or both.
+ */
+public interface ImmutableDisabledSlotsData
+		extends ImmutableDataManipulator<ImmutableDisabledSlotsData, DisabledSlotsData> {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/data/manipulator/immutable/entity/ImmutableDisabledSlotsData.java]: Can be one line.

File: src/main/java/org/spongepowered/api/data/manipulator/immutable/entity/ImmutableDisabledSlotsData.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+	/**
+	 * If players are prevented from taking items from each slot.
+	 * 
+	 * @return Whether each slot has taking disabled.
+	 */
+	ImmutableValue<Map<EquipmentType, Boolean>> takingDisabled();
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/data/manipulator/immutable/entity/ImmutableDisabledSlotsData.java]: `ImmutableMapValue`

File: src/main/java/org/spongepowered/api/data/manipulator/mutable/entity/DisabledSlotsData.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ * A slot can have taking disabled, placing disabled or both.
+ */
+public interface DisabledSlotsData extends DataManipulator<DisabledSlotsData, ImmutableDisabledSlotsData> {
+
+	/**
+	 * If players are prevented from taking items from each slot.
+	 * 
+	 * @return Whether each slot has taking disabled.
+	 */
+	Value<Map<EquipmentType, Boolean>> takingDisabled();
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/data/manipulator/mutable/entity/DisabledSlotsData.java]: `MapValue`.

**Review by gabizou — CHANGES_REQUESTED** [review]

**Comment by gabizou** [comment]
I appreciate the new additions! Definitely something that's been neglected. If possible, can you re-target this for `stable-7` so that it can be used in that API version? Do you intend to implement this? If you would like, we can provide some sort of assistance in understanding the implementation to do so.

File: src/main/java/org/spongepowered/api/data/key/Keys.java (hunk 2 above)
Comment (kashike) [reply src/main/java/org/spongepowered/api/data/key/Keys.java]: Or, alternatively, a `CollectionValue` - the `EquipmentType` simply being in the collection is enough, no need for a map-based solution here.

**Review by kashike — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/api/data/manipulator/immutable/entity/ImmutableDisabledSlotsData.java (hunk 4 above)
Comment (Lignium) [reply src/main/java/org/spongepowered/api/data/manipulator/immutable/entity/ImmutableDisabledSlotsData.java]: @gabizou This line is slightly longer than 120 characters.

**Comment by tcmal** [comment]
Moved to `stable-7` and used SetValue instead.

I can probably implement it.

File: src/main/java/org/spongepowered/api/data/manipulator/immutable/entity/ImmutableDisabledSlotsData.java (hunk 4 above)
Comment (ItsDoot) [reply src/main/java/org/spongepowered/api/data/manipulator/immutable/entity/ImmutableDisabledSlotsData.java]: @Lignium Sponge uses 150, not 120.

Comment (gabizou) [reply src/main/java/org/spongepowered/api/data/manipulator/immutable/entity/ImmutableDisabledSlotsData.java]: > This line is slightly longer than 120 characters.

Yeah... for the sake of simplicity, that "code style" hasn't been followed for years. Plenty of lines where that's not followed, but for class declarations like this, with a single extension, I'd say it's *fine to extend past the desired length*.

File: src/main/java/org/spongepowered/api/data/key/Keys.java [hunk 7]
```diff
@@ -277,6 +279,20 @@
      */
     public static final Key<Value<Boolean>> ARMOR_STAND_MARKER = DummyObjectProvider.createExtendedFor(Key.class,"ARMOR_STAND_MARKER");
 
+    /**
+     * Represents the {@link Key} for whether players are prevented from taking
```
Comment (gabizou) [inline src/main/java/org/spongepowered/api/data/key/Keys.java:283]:
```suggestion
     * Represents the {@link Key} for whether a {@link Player} can take an
     * {@link ItemStack} out of a specific {@link EquipmentSlot} slot from
     * an {@link ArmorStand}.
```

Comment (gabizou) [reply src/main/java/org/spongepowered/api/data/key/Keys.java:283]: Can swap out the other java doc as well, or propose something better? Just my two cents on the doc for clarity.

**Review by gabizou — COMMENTED** [review]
Probably want to add `default` methods to `ArmorStand` interface, much like the [existing default methods](https://github.com/SpongePowered/SpongeAPI/blob/1e966facc18ff0c5e9456b4ec0e9bf70f386633d/src/main/java/org/spongepowered/api/entity/living/ArmorStand.java#L44) so the data is more "visible" aside from the `Keys` class.

**Comment by gabizou** [comment]
> I can probably implement it.

Let me know if you've got questions, I'm available on our [discord](https://discord.gg/sponge) to point you to some example implementations and basically what to look out for.

Some immediate examples of an implementation that would closer match the "feature implementation" for this would be [the implementation of the ~~bloated~~ existing armor stand data for rotations](https://github.com/SpongePowered/SpongeCommon/commit/2f12890794ddd6bb2b9d33cf5602e4d57095969b).

**Pull request merged**